// @burner-wallet/modern-ui/dist/internal-plugins/balance-plugin/BalanceItem.js
import React, { useRef, useLayoutEffect } from 'react';
import styled from 'styled-components';
import { toBN } from 'web3-utils';
const SPEED = 4;
const BalanceCard = styled.div `
  display: inline-flex;
  vertical-align: middle;
  font-size: 48px;
  font-weight: 400;
  text-align: right;
  padding: 8px 16px 8px 8px;
  border: 1px solid #ccc;
  border-radius: 8px;
  min-width: 225px;
  background: white;

  &:not(:first-of-type) {
    margin-left: 12px;
  }
`;
const BalanceText = styled.div `
  flex: 1;
  display: flex;
  flex-direction: column;
  text-align: right;
`;
const Value = styled.div `
  font-size: 48px;
  display: flex;
  justify-content: flex-end;

  & .digit {
    width: 30px;
    text-align: center;
  }
`;
const AssetName = styled.div `
  font-size: 16px;
`;
const Icon = styled.div `
  min-width: 60px;
  background-position: center;
  background-repeat: no-repeat;
  background-size: contain;
`;
const getValue = (asset, balance) => {
    if (!balance) {
        return '-';
    }
    try {
        const usdVal = asset.getUSDValue(balance);
        return `$${parseFloat(usdVal).toFixed(2)}`;
    }
    catch (e) {
        const displayVal = asset.getDisplayValue(balance);
        return parseFloat(displayVal).toFixed(2);
    }
};
const BalanceItem = ({ asset, balance, growthRate }) => {
    const valueDiv = useRef(null);
    const value = getValue(asset, balance);
    useLayoutEffect(() => {
        if (!balance || growthRate === '0') {
            return;
        }
        const startTime = Date.now();
        let req;
        let frame = 0;
        const updateNum = () => {
            if (valueDiv.current && frame++ % SPEED === 0) {
                const timeDiff = (Date.now() - startTime).toString();
                const valueDiff = toBN(growthRate).mul(toBN(timeDiff)).div(toBN('1000'));
                const newBalance = toBN(balance).add(valueDiff).toString();
                const displayVal = getValue(asset, newBalance);
                valueDiv.current.innerHTML = Array.from(displayVal)
                    .map((char) => `<div ${char !== '.' ? 'class="digit"' : ''}>${char}</div>`)
                    .join('');
            }
            req = window.requestAnimationFrame(updateNum);
        };
        updateNum();
        return () => void window.cancelAnimationFrame(req);
    }, [balance, asset, growthRate]);
    return (React.createElement(BalanceCard, null,
        asset.icon && (React.createElement(Icon, { style: { backgroundImage: `url('${asset.icon}')` } })),
        React.createElement(BalanceText, null,
            React.createElement(Value, { ref: valueDiv }, Array.from(value).map((char, index) => (React.createElement("div", { className: char !== '.' ? 'digit' : '', key: index }, char)))),
            React.createElement(AssetName, null, asset.name))));
};
export default BalanceItem;
