// @burner-wallet/ui-core/dist/BurnerUICore.js
import React, { Component } from 'react';
import { BrowserRouter as Router } from 'react-router-dom';
import './i18n';
import dataProviders from './data-providers';
import BurnerProvider from './BurnerProvider';
import BurnerRouter from './BurnerRouter';
import Plugins from './Plugins';
const REQUIRED_PAGES = ['/', '/confirm', '/pk'];
export default class BurnerUICore extends Component {
    constructor(props) {
        super(props);
        const internalPlugins = this.getInternalPlugins();
        const _plugins = props.plugins ? [...internalPlugins, ...props.plugins] : internalPlugins;
        this.plugins = new Plugins(_plugins, this);
        this.state = {
            pluginData: this.plugins.getData(),
        };
        this._burnerComponents = {
            ...dataProviders,
            ...this.burnerComponents(),
        };
    }
    getInternalPlugins() {
        return [];
    }
    componentDidMount() {
        this.plugins.onDataChange(pluginData => this.setState({ pluginData }));
    }
    getCore() {
        return this.props.core;
    }
    getAssets() {
        return this.props.core.getAssets();
    }
    router() {
        return (React.createElement(BurnerRouter, { pluginData: this.state.pluginData, pages: this.getPages() }));
    }
    render() {
        return (React.createElement(Router, null,
            React.createElement(BurnerProvider, { core: this.props.core, pluginData: this.state.pluginData, burnerComponents: this._burnerComponents }, this.content())));
    }
}
BurnerUICore.defaultProps = {
    plugins: [],
};
